src/functions.c

```diff
@@ -549,10 +550,58 @@ static void fcallCommandGeneric(client *c, int ro) {
         return;
     }
 
+    if ((fi->f_flags & FUNCTION_FLAG_NO_CLUSTER) && server.cluster_enabled) {
+        addReplyError(c, "Can not run function on cluster, 'no-cluster' flag is set.");
+        return;
+    }
+
+    if ((fi->f_flags & FUNCTION_FLAG_DENY_OOM) && server.script_oom) {
+        addReplyError(c, "-OOM can not run the function");
+        return;
+    }
+
+    if (server.masterhost && server.repl_state != REPL_STATE_CONNECTED &&
+        server.repl_serve_stale_data == 0 && !(fi->f_flags & FUNCTION_FLAG_ALLOW_STALE))
+    {
+        addReplyError(c, "-MASTERDOWN Link with MASTER is down, "
+                         "replica-serve-stale-data is set to 'no' "
+                         "and 'allow-stale' flag is not set on the function.");
+        return;
+    }
+
+    if (fi->f_flags & FUNCTION_FLAG_WRITE) {
+        /* Function has the 'write' flag, make sure:
+         * 1. we are not a readonly replica
+         * 2. no disc error detected
+         * 3. command is not 'fcall_ro' */
+        if (server.masterhost && server.repl_slave_ro && c->id != CLIENT_ID_AOF
+            && !(c->flags & CLIENT_MASTER))
+        {
+            addReplyError(c, "Can not run a function with write flag on readonly replica");
+            return;
+        }
+
+        int deny_write_type = writeCommandsDeniedByDiskError();
+        if (deny_write_type != DISK_ERROR_TYPE_NONE && server.masterhost == NULL) {
+            if (deny_write_type == DISK_ERROR_TYPE_RDB)
+                addReplyError(c, shared.bgsaveerr->ptr);
+            else
+                addReplyErrorFormat(c, "-MISCONF Errors writing to the AOF file: %s", strerror(server.aof_last_write_errno));
```

Comment:
if only there was a nice way to share this error message code..
i guess what we have now is better than the alternatives,

Reply:
Changed it to a specific error message for function